imap_processing/hi/l1b/hi_l1b.py

```diff
@@ -90,28 +110,149 @@ def annotate_direct_events(l1a_dataset: xr.Dataset) -> xr.Dataset:
         "hae_latitude",
         "hae_longitude",
         "quality_flag",
         "nominal_bin",
     ]
     new_data_vars = create_dataset_variables(
-        l1b_de_var_names, (n_epoch,), att_manager_lookup_str="hi_de_{0}"
+        l1b_de_var_names, l1a_dataset["epoch"].size, att_manager_lookup_str="hi_de_{0}"
     )
-    l1b_dataset = l1a_dataset.assign(new_data_vars)
+    l1b_dataset = l1b_dataset.assign(new_data_vars)
     l1b_dataset = l1b_dataset.drop_vars(
         ["tof_1", "tof_2", "tof_3", "de_tag", "ccsds_met", "meta_event_met"]
     )
 
     de_global_attrs = ATTR_MGR.get_global_attributes("imap_hi_l1b_de_attrs")
     l1b_dataset.attrs.update(**de_global_attrs)
     return l1b_dataset
+
+
+def compute_coincidence_type_and_time_deltas(dataset: xr.Dataset) -> xr.Dataset:
+    """
+    Compute coincidence type and time deltas.
+
+    Adds the new variables "coincidence_type", "delta_t_ab", "delta_t_ac1",
+    "delta_t_bc1", and "delta_t_c1c2" to the input xarray.Dataset and returns
+    the updated xarray.Dataset.
+
+    Parameters
+    ----------
+    dataset : xarray.Dataset
+        The L1A/B dataset that results from reading in the L1A CDF and
+        allocating the new L1B DataArrays.
+
+    Returns
+    -------
+    xarray.Dataset
+        Updated xarray.Dataset with 5 new variables added.
+    """
+    new_data_vars = create_dataset_variables(
+        [
+            "coincidence_type",
+            "delta_t_ab",
+            "delta_t_ac1",
+            "delta_t_bc1",
+            "delta_t_c1c2",
+        ],
+        len(dataset.epoch),
+        "hi_de_{0}",
+    )
+    out_ds = dataset.assign(new_data_vars)
+
+    # compute masks needed for coincidence type and delta t calculations
+    a_mask = out_ds.trigger_id.values == TriggerId.A
+    b_mask = out_ds.trigger_id.values == TriggerId.B
+    c_mask = out_ds.trigger_id.values == TriggerId.C
+
+    tof_1_valid_mask = np.isin(
+        out_ds.tof_1.values, HiConstants.TOF1_BAD_VALUES, invert=True
+    )
+    tof_2_valid_mask = np.isin(
+        out_ds.tof_2.values, HiConstants.TOF2_BAD_VALUES, invert=True
+    )
+    tof_1and2_valid_mask = tof_1_valid_mask & tof_2_valid_mask
+    tof_3_valid_mask = np.isin(
+        out_ds.tof_3.values, HiConstants.TOF3_BAD_VALUES, invert=True
+    )
+
+    # Table denoting how hit-first mask and valid TOF masks are used to set
+    # coincidence type bitmask
+    # -----------------------------------------------------------------------
+    # | Trigger ID  |  Hit First  | TOF 1 Valid | TOF 2 Valid | TOF 3 Valid |
+    # -----------------------------------------------------------------------
+    # |      1      |      A      |     A,B     |     A,C1    |    C1,C2    |
+    # |      2      |      B      |     A,B     |     B,C1    |    C1,C2    |
+    # |      3      |      C1     |     A,C1    |     B,C1    |    C1,C2    |
+    # Set coincidence type bitmask
+    out_ds.coincidence_type[a_mask | tof_1_valid_mask] |= CoincidenceBitmap.A
```

Comment:
For me personally I would really appreciate a docstring on each bitmask. Maybe it's overkill for some people, but it helps me to immediately understand the purpose.

Example:
Set the 'A' bit in the 'coincidence_type' variable where:
- The trigger is 'A' (using `a_mask`), OR
- `tof_1` is valid (using `tof_1_valid_mask`).

Reply:
My philosophy is that if you need a bunch of comments to explain your code, then the code is not written in a way that it can be clearly understood. IMO, the comments you wrote are exactly what the code says with a slightly different order. For this instance, I think the comment is redundant.